1306_expansion_statements/expansion-statements.md

```diff
@@ -469,6 +469,409 @@ void f() {
+
+::: std
+::: addu
+**Expansion statements   [stmt.expand]**
+
+[1]{.pnum} Expansion statements specify repeated instantiations ([temp.spec]) of their substatement.
+
+```
+$expansion-statement$:
+    template for ( $init-statement$@~_opt_~@ $for-range-declaration$ : $expansion-initializer$ ) $statement$
+
+$expansion-initializer$:
+    $expression$
+    $expansion-init-list$
+
+$expansion-init-list$:
+    { $expression-list$ }
+```
+
+[Each `$decl-specifier$` in the `$decl-specifier-seq$` of a `$for-range-declaration$` is either a `$type-specifier$` or `constexpr` ([stmt.ranged]).]{.note}
+
+[#]{.pnum} The `$statement$` of an `$expansion-statement$` is a control-flow-limited statement ([stmt.label]).
+
+[#]{.pnum} For the purpose of name lookup and instantiation, the `$for-range-declaration$` and the `$statement$` of the `$expansion-statement$` are together considered a template definition.
+
+[#]{.pnum} An expansion statement whose `$expansion-initializer$` is an iterable expression ([stmt.ranged]) that does not have array type is an _iterable expansion statement_. An expansion statement that is not iterable and whose `$expansion-initializer$` is a non-type-dependent expression with a structured binding size ([dcl.struct.bind]) is a _destructurable expansion statement_. An expansion statement that is neither iterable nor destructurable shall have an `$expansion-initializer$` of the form `$expansion-init-list$` and is an _enumerated expansion statement_.
+
+[#]{.pnum} An expansion statement `$S$` with an `$expansion-initializer$` `$E$` is _expanded_ unless either
+
+- [#.#]{.pnum} `$E$` is type-dependent,
+- [#.#]{.pnum} `$S$` is an iterable expansion statement and `$E$` is value-dependent ([temp.dep.constexpr]) or
+- [#.#]{.pnum} `$S$` is an enumerated expansion statement and `$E$` contains a pack.
+
+[#]{.pnum} An expanded expansion statement `$S$` is equivalent to a `$compound-statement$` containing instantiations of the `$for-range-declaration$` (including its implied initialization), together with the `$statement$`; these instantiations correspond to expressions resulting from the analysis of the `$expansion-initializer$` as follows:
+
+- [#.#]{.pnum} If `$S$` is an iterable expansion statement, there is an instantiation for each element in the range computed by the `$expansion-initializer$`; `$S$` is equivalent to:
+  ```cpp
+  {
+    $init-statement$
+    static constexpr auto&& range = $expansion-initializer$ ;
+    // @i^th^@ repetition of the substatement
+    static constexpr auto iter@~_i_~@ = $get-expr$@~_i_~@ ;
+    {
+      $for-range-declaration$ = *iter@~_i_~@ ;
+      $statement$
+    }
+  }
+  ```
+
+  where `$get-expr$@~_0_~@` is the expression:
+  ```cpp
+  begin($range$)
+  ```
+
+  and every subsequent `$get-expr$@~_i_~@` is the expression:
+  ```cpp
+  $get-expr$@~_i-1_~@ + 1
+  ```
+
+  for all `i` in the range [0, _N_], for _N_ such that:
+  ```cpp
+  $get-expr$@~_N_~@ == end($range$) @.@
```

Comment:
So basically we just count up until we get there, okay that we don't need N up front? 

Regardless, this should be end-expr.